### PR TITLE
Update RankingTrainer usage and Remove BLAS

### PR DESCRIPTION
**Pull request by OctoberChang**
*Issue #, if available:*

*Description of changes:*
- Slight adjustment of RankingTrainer class to support custom loss functions for its child-class
- Remove dependency of BLAS, so the package is more lightweight
- Bump minimal support Python version from 3.8 to 3.9 (Py3.8 is end-of-life as of 2024/10/07)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/build_pypi_wheel.sh [hunk 1]
```diff
@@ -19,25 +19,8 @@ echo "pip: $($PIP --version)"
...
-   OPENBLAS_LIB="openblas-${OPENBLAS_VER}-${PLAT}.tar.gz"
-   OPENBLAS_LIB_URL="https://anaconda.org/multibuild-wheels-staging/openblas-libs/$OPENBLAS_VER/download/$OPENBLAS_LIB"
-   yum install wget -y
-   wget $OPENBLAS_LIB_URL
-   tar -xvf $OPENBLAS_LIB
-else
-   echo "$PLAT not supported."
-   exit 1
-fi
+$PIP install 'setuptools' wheel twine auditwheel
```
Comment (rofuyu) [inline .github/build_pypi_wheel.sh]: do we still need the single quotes here?

File: pecos/core/utils/matrix.hpp [hunk 2]
```diff
@@ -764,37 +764,46 @@ namespace pecos {
...
-    template<typename val_type> val_type dot(ptrdiff_t *, val_type *, ptrdiff_t *, val_type *, ptrdiff_t *);
-    template<> inline double dot(ptrdiff_t *len, double *x, ptrdiff_t *xinc, double *y, ptrdiff_t *yinc) { return ddot_(len, x, xinc, y, yinc); }
-    template<> inline float dot(ptrdiff_t *len, float *x, ptrdiff_t *xinc, float *y, ptrdiff_t *yinc) { return sdot_(len, x, xinc, y, yinc); }
-
-    template<typename val_type> val_type scal(ptrdiff_t *, val_type *, val_type *, ptrdiff_t *);
-    template<> inline double scal(ptrdiff_t *len, double *a, double *x, ptrdiff_t *xinc) { return dscal_(len, a, x, xinc); }
-    template<> inline float scal(ptrdiff_t *len, float *a,  float *x, ptrdiff_t *xinc) { return sscal_(len, a, x, xinc); }
+    template<typename val_type> val_type scal(ptrdiff_t *len, val_type *a, val_type *x, ptrdiff_t *xinc) {
+	for (ptrdiff_t idx = 0; idx < *len; idx++) {
+	    *x = (*x) * (*a);
```
Comment (rofuyu) [inline pecos/core/utils/matrix.hpp]: indentation space vs tab?

File: setup.py [hunk 3]
```diff
@@ -140,10 +118,9 @@ def get_blas_lib_dir(cls):
     "pecos.core.libpecos_float32",
     sources=["pecos/core/libpecos.cpp"],
     include_dirs=["pecos/core", "/usr/include/", "/usr/local/include"],
-    libraries=["gomp", "gcc"] + blas_lib,
-    library_dirs=blas_dir,
+    libraries=["gomp", "gcc", "stdc++"],
     extra_compile_args=["-fopenmp", "-O3", "-std=c++17"] + manual_compile_args,
-    extra_link_args=['-Wl,--no-as-needed', f"-Wl,-rpath,{':'.join(blas_dir)}"]
+    extra_link_args=['-Wl,--no-as-needed', f"-Wl,-rpath"]
```
Comment (rofuyu) [inline setup.py]: we might be able to remove the rpath path.

Comment (OctoberChang) [reply setup.py]: Thanks for the suggestion. We removed the `extra_link_args` in the latest commit.

File: pecos/core/utils/matrix.hpp (hunk 2 above)
Comment (OctoberChang) [reply pecos/core/utils/matrix.hpp]: Thanks for the catch. Fixed the indentation issue in the latest commit.

File: .github/build_pypi_wheel.sh (hunk 1 above)
Comment (OctoberChang) [reply .github/build_pypi_wheel.sh]: Thanks for the catch. Removed the single quotes in the latest commit.

**Review by rofuyu — APPROVED** [review]

**Pull request merged**